File: apps/web/lib/api/evaluation.ts
import { dashboardApiGet } from "./client";
import type { DevelopmentDashboardSession } from "../auth/development-session";

export type EvaluationDataset = {
  id: string;
  organisation_id: string;
  workspace_id: string;
  widget_id: string;
  name: string;
  description: string | null;
  version: string;
  status: string;
  created_by: string | null;
  created_at: string;
  updated_at: string;
};

export type EvaluationCase = {
  id: string;
  dataset_id: string;
  question: string;
  reference_answer: string | null;
  expected_document_ids: string[] | null;
  expected_source_labels: string[] | null;
  expected_answerability: string;
  category: string;
  tags: string[] | null;
  metadata_json: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type EvaluationDatasetDetail = EvaluationDataset & { cases: EvaluationCase[] };

export type EvaluationRun = {
  id: string;
  organisation_id: string;
  workspace_id: string;
  widget_id: string;
  dataset_id: string;
  dataset_version: string;
  provider_key: string | null;
  model_key: string | null;
  provider_model_name: string | null;
  prompt_key: string | null;
  prompt_version: string | null;
  prompt_hash: string | null;
  mode: string;
  status: string;
  total_cases: number;
  passed_cases: number;
  failed_cases: number;
  hard_failure_cases: number;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
};

export type EvaluationCategoryBreakdown = Record<string, { total: number; passed: number; failed: number; hard_failure: number }>;

export type EvaluationFailedCase = {
  case_id: string;
  question: string;
  category: string;
  passed: boolean;
  hard_failure: boolean;
  failure_reasons: string[];
  latency_ms: number | null;
};

export type EvaluationRunSummary = {
  total_cases: number;
  passed_cases: number;
  failed_cases: number;
  hard_failure_cases: number;
  pass_rate: number;
  retrieval_hit_rate: number | null;
  citation_coverage: number;
  fallback_rate_on_answerable: number | null;
  correct_fallback_rate_on_unanswerable: number | null;
  latency_p50_ms: number | null;
  latency_p95_ms: number | null;
  total_prompt_tokens: number;
  total_completion_tokens: number;
  total_tokens: number;
  category_breakdown: EvaluationCategoryBreakdown;
  failed_case_details: EvaluationFailedCase[];
};

export type EvaluationGateVerdict = { passed: boolean; reasons: string[] };

export type EvaluationRunDetail = {
  run: EvaluationRun;
  summary: EvaluationRunSummary;
  gate: EvaluationGateVerdict;
};

export type EvaluationResult = {
  id: string;
  run_id: string;
  case_id: string;
  actual_answer: string | null;
  answer_state: string | null;
  retrieved_document_ids: string[] | null;
  retrieved_chunk_ids: string[] | null;
  citations_json: Array<Record<string, unknown>> | null;
  latency_ms: number | null;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  total_tokens: number | null;
  retrieval_metrics_json: Record<string, unknown> | null;
  answer_metrics_json: Record<string, unknown> | null;
  judge_scores_json: Record<string, unknown> | null;
  passed: boolean;
  hard_failure: boolean;
  failure_reasons_json: string[] | null;
  error_message: string | null;
  created_at: string;
};

export type EvaluationResultDetail = { result: EvaluationResult; case: EvaluationCase | null };

export type EvaluationRunComparison = {
  baseline: { run: EvaluationRun; summary: EvaluationRunSummary };
  candidate: { run: EvaluationRun; summary: EvaluationRunSummary };
  comparison: {
    baseline_pass_rate: number;
    candidate_pass_rate: number;
    pass_rate_delta: number;
    baseline_hard_failure_cases: number;
    candidate_hard_failure_cases: number;
    baseline_latency_p95_ms: number | null;
    candidate_latency_p95_ms: number | null;
    regressed: boolean;
  };
};

function workspacePath(session: DevelopmentDashboardSession, suffix: string) {
  return `/api/v1/workspaces/${session.workspaceId}${suffix}`;
}

function tenantParams(session: DevelopmentDashboardSession) {
  return { organisation_id: session.organisationId };
}

export function listEvaluationDatasets(session: DevelopmentDashboardSession) {
  return dashboardApiGet<EvaluationDataset[]>({
    path: workspacePath(session, "/evaluation/datasets"),
    session,
    searchParams: tenantParams(session),
  });
}

export function getEvaluationDataset(session: DevelopmentDashboardSession, datasetId: string) {
  return dashboardApiGet<EvaluationDatasetDetail>({
    path: workspacePath(session, `/evaluation/datasets/${datasetId}`),
    session,
    searchParams: tenantParams(session),
  });
}

export function listEvaluationRuns(session: DevelopmentDashboardSession, datasetId?: string) {
  return dashboardApiGet<EvaluationRun[]>({
    path: workspacePath(session, "/evaluation/runs"),
    session,
    searchParams: { ...tenantParams(session), ...(datasetId ? { dataset_id: datasetId } : {}) },
  });
}

export function getEvaluationRun(session: DevelopmentDashboardSession, runId: string) {
  return dashboardApiGet<EvaluationRunDetail>({
    path: workspacePath(session, `/evaluation/runs/${runId}`),
    session,
    searchParams: tenantParams(session),
  });
}

export function listEvaluationRunResults(session: DevelopmentDashboardSession, runId: string, onlyFailed?: boolean) {
  return dashboardApiGet<EvaluationResult[]>({
    path: workspacePath(session, `/evaluation/runs/${runId}/results`),
    session,
    searchParams: { ...tenantParams(session), ...(onlyFailed ? { only_failed: "true" } : {}) },
  });
}

export function getEvaluationRunResult(session: DevelopmentDashboardSession, runId: string, caseId: string) {
  return dashboardApiGet<EvaluationResultDetail>({
    path: workspacePath(session, `/evaluation/runs/${runId}/results/${caseId}`),
    session,
    searchParams: tenantParams(session),
  });
}

export function compareEvaluationRuns(session: DevelopmentDashboardSession, baselineRunId: string, candidateRunId: string) {
  return dashboardApiGet<EvaluationRunComparison>({
    path: workspacePath(session, "/evaluation/runs/compare"),
    session,
    searchParams: { ...tenantParams(session), baseline_run_id: baselineRunId, candidate_run_id: candidateRunId },
  });
}

File: apps/api/app/schemas/evaluation.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.evaluation.categories import ANSWERABILITY_VALUES, CASE_CATEGORY_VALUES


class EvaluationDatasetCreate(BaseModel):
    widget_id: str
    name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=4000)
    version: str = Field(default="1", min_length=1, max_length=40)


class EvaluationDatasetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    organisation_id: str
    workspace_id: str
    widget_id: str
    name: str
    description: str | None
    version: str
    status: str
    created_by: str | None
    created_at: datetime
    updated_at: datetime


class EvaluationCaseCreate(BaseModel):
    question: str = Field(min_length=1)
    reference_answer: str | None = None
    expected_document_ids: list[str] | None = None
    expected_source_labels: list[str] | None = None
    expected_answerability: str = "answerable"
    category: str
    tags: list[str] | None = None
    metadata_json: dict | None = None

    def validate_vocabulary(self) -> None:
        if self.expected_answerability not in ANSWERABILITY_VALUES:
            raise ValueError(f"expected_answerability must be one of {sorted(ANSWERABILITY_VALUES)}")
        if self.category not in CASE_CATEGORY_VALUES:
            raise ValueError(f"category must be one of {sorted(CASE_CATEGORY_VALUES)}")


class EvaluationCaseUpdate(BaseModel):
    question: str | None = Field(default=None, min_length=1)
    reference_answer: str | None = None
    expected_document_ids: list[str] | None = None
    expected_source_labels: list[str] | None = None
    expected_answerability: str | None = None
    category: str | None = None
    tags: list[str] | None = None
    metadata_json: dict | None = None


class EvaluationCaseRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    dataset_id: str
    question: str
    reference_answer: str | None
    expected_document_ids: list[str] | None
    expected_source_labels: list[str] | None
    expected_answerability: str
    category: str
    tags: list[str] | None
    metadata_json: dict | None
    created_at: datetime
    updated_at: datetime


class EvaluationDatasetDetail(EvaluationDatasetRead):
    cases: list[EvaluationCaseRead]


class EvaluationRunCreate(BaseModel):
    dataset_id: str
    widget_id: str
    mode: str = "mock"


class EvaluationRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    organisation_id: str
    workspace_id: str
    widget_id: str
    dataset_id: str
    dataset_version: str
    provider_key: str | None
    model_key: str | None
    provider_model_name: str | None
    prompt_key: str | None
    prompt_version: str | None
    prompt_hash: str | None
    mode: str
    status: str
    total_cases: int
    passed_cases: int
    failed_cases: int
    hard_failure_cases: int
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    trigger_source: str | None = None


class EvaluationResultRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    run_id: str
    case_id: str
    actual_answer: str | None
    answer_state: str | None
    retrieved_document_ids: list[str] | None
    retrieved_chunk_ids: list[str] | None
    citations_json: list[dict] | None
    latency_ms: int | None
    prompt_tokens: int | None
    completion_tokens: int | None
    total_tokens: int | None
    retrieval_metrics_json: dict | None
    answer_metrics_json: dict | None
    judge_scores_json: dict | None
    passed: bool
    hard_failure: bool
    failure_reasons_json: list[str] | None
    error_message: str | None
    created_at: datetime
